src/control/ras/event.go

```diff
@@ -0,0 +1,205 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package ras
+
+/*
+#include "daos_srv/ras.h"
+*/
+import "C"
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/common"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+// EventID describes a given RAS event.
+type EventID uint32
+
+// EventID constant definitions.
+const (
+	EventRankFail   EventID = C.RAS_RANK_FAIL
+	EventRankNoResp EventID = C.RAS_RANK_NO_RESP
+)
+
+func (id EventID) String() string {
+	return C.GoString(C.ras_event_id_enum_to_name(uint32(id)))
+}
+
+// Desc returns a description of the event.
+func (id EventID) Desc() string {
+	return C.GoString(C.ras_event_id_enum_to_msg(uint32(id)))
+}
+
+// Uint32 returns uint32 representation of event ID.
+func (id EventID) Uint32() uint32 {
+	return uint32(id)
+}
+
+// EventSeverityID describes the severity of a given RAS event.
+type EventSeverityID uint32
+
+// EventSeverityID constant definitions.
+const (
+	EventSeverityFatal EventSeverityID = C.RAS_SEV_FATAL
+	EventSeverityWarn  EventSeverityID = C.RAS_SEV_WARN
+	EventSeverityError EventSeverityID = C.RAS_SEV_ERROR
+	EventSeverityInfo  EventSeverityID = C.RAS_SEV_INFO
+)
+
+func (sev EventSeverityID) String() string {
+	return C.GoString(C.ras_event_sev_enum_to_name(uint32(sev)))
+}
+
+// Uint32 returns uint32 representation of event severity.
+func (sev EventSeverityID) Uint32() uint32 {
+	return uint32(sev)
+}
+
+// EventTypeID describes the type of a given RAS event.
+type EventTypeID uint32
+
+// EventTypeID constant definitions.
+const (
+	EventTypeStateChange EventTypeID = C.RAS_TYPE_STATE_CHANGE
+	EventTypeInfoOnly    EventTypeID = C.RAS_TYPE_INFO_ONLY
+)
+
+func (typ EventTypeID) String() string {
+	return C.GoString(C.ras_event_type_enum_to_name(uint32(typ)))
+}
+
+// Uint32 returns uint32 representation of event type.
+func (typ EventTypeID) Uint32() uint32 {
+	return uint32(typ)
+}
+
+// Event describes details of a specific RAS event.
+type Event struct {
+	Name        string `json:"name"`
+	Timestamp   string `json:"timestamp"`
+	Msg         string `json:"msg"`
+	Hostname    string `json:"hostname"`
+	Data        []byte `json:"data"`
+	Rank        uint32 `json:"rank"`
+	InstanceIdx uint32 `json:"instance_idx"`
+	ID          EventID
```

Comment:
but they aren't converted directly, they have to be handled specifically in the Un/Marshal methods for the type, in that case , should they still have the Json tags?

Reply:
Yes, because these tags aren't just for conversion between types. They're also used when the type is serialized to JSON for external consumers (e.g. dmg --json, etc).

Reply:
done

Reply:
adding the json tags for the typed fields breaks conversion from proto format to native because the custom Marshal/Unmarshal methods are not defined on the auto generated proto types. I think the tags are not necessary because when outputting json the Marshal/Unmarshal methods will handle the conversion without the tags

Reply:
not specifying the tags for types that have custom Marshal/Unmarshal    matches what we have been doing previously